File: 20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */
#include <stack>
#include <string>

using std::stack;
using std::string;

// @lc code=start
class Solution {
public:
    // stack
    // T: O(N)
    // S: O(N)
    bool isValid(string s) {
        stack<char> stk;
        for (const auto &c : s) {
            if (c == '(' || c == '[' || c == '{') {
                stk.push(c);
            } else {
                if (stk.empty())  return false;
                auto sc = stk.top();
                if ((c == ')' && sc != '(') || 
                    (c == ']' && sc != '[') || 
                    (c == '}' && sc != '{'))
                    return false;
                stk.pop();
            }       
        }
        return stk.empty();
    }
};
// @lc code=end

File: 20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */
import java.util.Stack;

// @lc code=start
class Solution {
    // T: O(n)
    // S: O(n)
    public boolean isValid(String s) {
        Stack<Character> stk = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') {
                stk.push(c);
            } else {
                if (stk.isEmpty()) return false;
                if (c == ')' && stk.peek() != '(') return false;
                if (c == ']' && stk.peek() != '[') return false;
                if (c == '}' && stk.peek() != '{') return false;
                stk.pop();
            }
        }
        return stk.isEmpty();
    }
}
// @lc code=end

